presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/StatisticsUtil.java

```diff
@@ -57,29 +68,49 @@ public static TableStatistics mergeHiveStatistics(TableStatistics icebergStatist
-                    hiveColumnStats.getNullsCount().ifPresent(nullCount -> {
-                        Estimate nullsFraction;
-                        if (!hiveStatistics.getBasicStatistics().getRowCount().isPresent()) {
-                            if (icebergStatistics.getRowCount().isUnknown()) {
-                                nullsFraction = Estimate.unknown();
-                            }
-                            else {
-                                nullsFraction = Estimate.of((double) nullCount / icebergStatistics.getRowCount().getValue());
-                            }
-                        }
-                        else {
-                            nullsFraction = Estimate.of((double) nullCount / hiveStatistics.getBasicStatistics().getRowCount().getAsLong());
-                        }
-                        mergedStats.setNullsFraction(nullsFraction);
-                    });
+                // data size
+                if (mergeFlags.contains(ColumnStatisticType.TOTAL_SIZE_IN_BYTES)) {
+                    hiveColumnStats.getTotalSizeInBytes().ifPresent(size -> mergedStats.setDataSize(Estimate.of(size)));
                 }
             }
             statsBuilder.setColumnStatistics(columnHandle, mergedStats.build());
         });
-        return statsBuilder.build();
+        return calculateAndSetTableSize(statsBuilder).build();
+    }
+
+    public static EnumSet<ColumnStatisticType> decodeMergeFlags(String input)
+    {
+        return Optional.of(Arrays.stream((input).trim().split(","))
+                        .filter(value -> !value.isEmpty())
+                        .map(ColumnStatisticType::valueOf)
+                        .collect(Collectors.toSet()))
+                .filter(set -> !set.isEmpty())
+                .map(EnumSet::copyOf)
+                .orElse(EnumSet.noneOf(ColumnStatisticType.class));
+    }
+
+    public static String encodeMergeFlags(EnumSet<ColumnStatisticType> flags)
+    {
+        return Joiner.on(",").join(flags.stream().map(Enum::name).iterator());
+    }
+
+    public static TableStatistics.Builder calculateAndSetTableSize(TableStatistics.Builder builder)
+    {
+        return builder.setTotalSize(builder.getRowCount().flatMap(rowCount -> builder.getColumnStatistics().entrySet().stream().map(entry -> {
+            IcebergColumnHandle columnHandle = (IcebergColumnHandle) entry.getKey();
+            ColumnStatistics stats = entry.getValue();
+            return stats.getDataSize().or(() -> {
+                if (columnHandle.getType() instanceof FixedWidthType) {
+                    return stats.getNullsFraction().map(nulls -> rowCount * (1 - nulls) * ((FixedWidthType) columnHandle.getType()).getFixedSize());
+                }
+                else {
+                    return Estimate.unknown();
+                }
+            });
+        }).reduce(Estimate.of(0.0), (currentSize, newSize) -> currentSize.flatMap(current -> newSize.map(add -> current + add)))));
```

Comment:
Should we filter out metadata columns here or some place before here? Or we would always get an unknown estimate for the table's total size, as data size statistics for metadata columns of `VariableWidthType` would be unknown.

In addition, should we set total size to the row for whole table statistics in `ShowStatsRewrite.createTableStatsRow(..)`, so that we can read and get it through `show stats for...`?

Reply:
> Should we filter out metadata columns here or some place before here?

The calculation here only uses column statistics if they are added to the `TableStatistics.Builder`. As far as I know, the metadata columns (`SYNTHESIZED` from `BaseHiveColumnHandle`) are always filtered out before being added to the `TableStatistics.Builder`.

If data size of a `VariableWidthType` is unknown and it has column statistics added to `TableStatistics.Builder`, then it should follow that the total size of the table is unknown.


> In addition, should we set total size to the row for whole table statistics in ShowStatsRewrite.createTableStatsRow(..), so that we can read and get it through show stats for...?

I think this can be left to a separate PR? It would apply to all connectors and I feel is outside the scope of this PR since it's not Iceberg related. I think it should be added though. I will create a follow-up PR later.

Reply:
> The calculation here only uses column statistics if they are added to the `TableStatistics.Builder`. As far as I know, the metadata columns (`SYNTHESIZED` from `BaseHiveColumnHandle`) are always filtered out before being added to the `TableStatistics.Builder`.

Can you make sure of this? I think I just met this problem by simply executing a `show stats for <table_name>`. I see the metadata columns are passed in as `selectedColumns` for `TableStatisticsMaker.getTableStatistics()`.

Reply:
> > The calculation here only uses column statistics if they are added to the `TableStatistics.Builder`. As far as I know, the metadata columns (`SYNTHESIZED` from `BaseHiveColumnHandle`) are always filtered out before being added to the `TableStatistics.Builder`.
> 
> Can you make sure of this? I think I just met this problem by simply executing a `show stats for <table_name>`. I see the metadata columns are passed in as `selectedColumns` for `TableStatisticsMaker.getTableStatistics()`.

Ah you're right. It seems `SHOW STATS` passes the hidden columns. However, when running queries, unless the query uses that column it won't be passed. I think for the purpose of `SHOW STATS` that we can omit passing hidden columns to the `selectedColumns` list when running `getTableStatistics`, since they are filtered out later anyways.